src/legacy/core_plugins/embeddable_api/public/containers/container.ts

```diff
@@ -0,0 +1,381 @@
+    TEmbeddable extends IEmbeddable<TEmbeddableInput> = IEmbeddable<TEmbeddableInput>
+  >(type: string, savedObjectId: string): Promise<TEmbeddable | ErrorEmbeddable> {
+    const factory = this.embeddableFactories.get(type) as EmbeddableFactory<
+      TEmbeddableInput,
+      any,
+      TEmbeddable
+    >;
+    const panelState = this.createNewPanelState(factory);
+    panelState.savedObjectId = savedObjectId;
+
+    this.updateInput({
+      panels: {
+        ...this.input.panels,
+        [panelState.embeddableId]: panelState,
+      },
+    } as Partial<TContainerInput>);
+
+    return await this.untilEmbeddableLoaded<TEmbeddable>(panelState.embeddableId);
+  }
+
+  public removeEmbeddable(embeddableId: string) {
+    // Just a shortcut for removing the panel from input state, all internal state will get cleaned up naturally
+    // by the listener.
+    const panels = { ...this.input.panels };
+    delete panels[embeddableId];
+    this.updateInput({ panels } as Partial<TContainerInput>);
+  }
+
+  public getChild<E extends IEmbeddable>(id: string): E {
+    return this.children[id] as E;
+  }
+
+  public getInputForChild<TEmbeddableInput extends EmbeddableInput = EmbeddableInput>(
+    embeddableId: string
+  ): TEmbeddableInput {
+    const containerInput: TChildInput = this.getInheritedInput(embeddableId);
+    const panelState = this.getPanelState(embeddableId);
+
+    const explicitInput = panelState.explicitInput;
+    const explicitFiltered: { [key: string]: unknown } = {};
+
+    const keys = getKeys(panelState.explicitInput);
+
+    // If explicit input for a particular value is undefined, and container has that input defined,
+    // we will use the inherited container input. This way children can set a value to undefined in order
+    // to default back to inherited input. However, if the particular value is not part of the container, then
+    // the caller may be trying to explicitly tell the child to clear out a given value, so in that case, we want
+    // to pass it along.
+    keys.forEach(key => {
+      if (explicitInput[key] === undefined && containerInput[key] !== undefined) {
+        return;
+      }
+      explicitFiltered[key] = explicitInput[key];
+    });
+
+    return ({
+      ...containerInput,
+      ...explicitFiltered,
+      // Typescript has difficulties with inferring this type but it is accurate with all
+      // tests I tried. Could probably be revisted with future releases of TS to see if
+      // it can accurately infer the type.
+    } as unknown) as TEmbeddableInput;
+  }
+
+  public destroy() {
+    super.destroy();
+    Object.values(this.children).forEach(child => child.destroy());
+    this.subscription.unsubscribe();
+  }
+
+  public async untilEmbeddableLoaded<TEmbeddable extends IEmbeddable>(
+    id: string
+  ): Promise<TEmbeddable | ErrorEmbeddable> {
+    return new Promise(resolve => {
+      const subscription = this.getOutput$().subscribe(() => {
```

Comment:
output of all _container_ embeddables will change when a child has finished loading. 

> is this working with every embeddable type ? 

all container types, because it's handled in the base class.

> can we also use that outside of containers then ?

I'm not quite sure what you mean.  it's public so someone who has the container object can use `myContainer.untilEmbeddableLoaded` Which if that is what you meant, now that you mention it, i could use this logic inside embeddable_child_panel instead of manually creating the subscription.  I'll update that.  And I'll add tests because I just realized a bug if the panel is removed before it finished loading, it can possibly never return.

Reply:
if i use embeddable directly in my app (discover), am i able to check when it has loaded ?